core/src/main/java/org/apache/druid/java/util/metrics/AllocationMetricCollector.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.java.util.metrics;
+
+import it.unimi.dsi.fastutil.longs.Long2LongOpenHashMap;
+import org.apache.druid.java.util.common.logger.Logger;
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+class AllocationMetricCollector
+{
+  private static final Logger log = new Logger(AllocationMetricCollector.class);
+  private static Method getThreadAllocatedBytes;
+  private static ThreadMXBean threadMXBean;
+
+  private Map<Long, Long> previousResults;
+  private static boolean initialized = false;
+
+  static {
+    try {
+      // classes in the sun.* packages are not part of the public/supported Java API and should not be used directly.
+      threadMXBean = ManagementFactory.getThreadMXBean();
+      getThreadAllocatedBytes = threadMXBean.getClass().getMethod("getThreadAllocatedBytes", long[].class);
+      getThreadAllocatedBytes.setAccessible(true);
+      getThreadAllocatedBytes.invoke(threadMXBean, (Object) threadMXBean.getAllThreadIds());
+      initialized = true;
+    }
+    catch (Exception e) {
+      log.warn(e, "Cannot initialize %s", AllocationMetricCollector.class.getName());
+    }
+  }
+
+  AllocationMetricCollector()
+  {
+    try {
+      if (initialized) {
+        previousResults = new Long2LongOpenHashMap();
+      }
+    }
+    catch (Exception e) {
+      log.warn(e, "Cannot initialize %s", AllocationMetricCollector.class.getName());
+    }
+  }
+
+  public Optional<Long> calculateDelta()
```

Comment:
Logically, once `AllocationMetricCollector` is created, it should always report the metric, because the risk are inaccessible APIs. So suggested to create a static factory that returns an instance of AllocationMetricCollector if successful, or none if not successful. and then `calculateDelta()` always returns just `long`.

Reply:
I guess then you have to check if the factory returns a true collector, still an additional check anyway.

Reply:
Yes, but it's coherent from type system perspective - because currently there is a method that returns `Optional<Long>`, it may seem that it may return empty or full from time to time, but in fact it always returns only one type of thing.